src/documents/models.py

```diff
@@ -267,17 +277,111 @@ def __str__(self):
         return str(created)
 
     @property
-    def source_path(self):
-
-        file_name = "{:07}.{}".format(self.pk, self.file_type)
+    def source_filename(self):
+        if self.filename is None:
+            self.filename = self.generate_source_filename()
+
+        return self.filename
+
+    @staticmethod
+    def many_to_dictionary(field):
+        # Converts ManyToManyField to dictionary by assuming, that field
+        # entries contain an _ or - which will be used as a delimiter
+        mydictionary = dict()
+
+        for t in field.all():
+            # Find delimiter
+            delimiter = t.name.find('_')
+
+            if delimiter is -1:
+                delimiter = t.name.find('-')
+
+            if delimiter is -1:
+                continue
+
+            key = t.name[:delimiter]
+            value = t.name[delimiter+1:]
+
+            mydictionary[slugify(key)] = slugify(value)
+
+        return mydictionary
+
+    @staticmethod
+    def many_to_list(field):
+        # Converts ManyToManyField to list
+        mylist = list()
+
+        for t in field.all():
+            mylist.append(slugify(t.name))
+
+        return mylist
+
+    @staticmethod
+    def fill_list(input_list, length, filler):
+        while len(input_list) < length:
+            input_list.append(slugify(filler))
+
+        return input_list
+
+    def generate_source_filename(self):
+        # Create filename based on configured format
+        if settings.PAPERLESS_FILENAME_FORMAT is not None:
+            tag = defaultdict(lambda: slugify(None),
+                              self.many_to_dictionary(self.tags))
+            list_length = 10
+            tags = self.many_to_list(self.tags)
+            while True:
+                tags = Document.fill_list(tags, list_length, None)
```

Comment:
Hm, I don't think I like this solution. The while-loops both here and in `fill_list` are pretty confusing if you read this code without prior knowledge.

A much simpler solution would be to "mis-use" a defaultdict and get this down to one line:

```python
tags = defaultdict(lambda: slugify(None),
                   enumerate(self.many_to_list(self.tags)))
```

This will create a dict indexed using the original list's indices. If an invalid index is provided, the default will be returned (`"none"`). (It does allow retrieval using non-integer indices, but the user shouldn't be doing that. Anyway, it will not cause errors, just return `"none"` as well.)

Doing it like that removes `fill_list` again and gets rid of the while-loop. In addition, the try-except can be removed as well.

Reply:
As an additional idea: how about the distinction between `tag` and `tags` is removed altogether, and we include both string and int indices in the defaultdict? I thought it could work like this:

```python
@staticmethod
def many_to_dictionary(field):
    # Converts ManyToManyField to dictionary by assuming, that field
    # entries contain an _ or - which will be used as a delimiter
    mydictionary = dict()

    for index, t in enumerate(field.all()):
        # Populate tag names by index
        mydictionary[index] = slugify(t.name)

        # Find delimiter
        delimiter = t.name.find('_')

        if delimiter is -1:
            delimiter = t.name.find('-')

        if delimiter is -1:
            continue

        key = t.name[:delimiter]
        value = t.name[delimiter+1:]

        mydictionary[slugify(key)] = slugify(value)

    return mydictionary

def generate_source_filename(self):
    if settings.PAPERLESS_FILENAME_FORMAT is not None:
        tags = defaultdict(lambda: slugify(None),
                           self.many_to_dictionary(self.tags))
        # ...
```

Then it could be used like this:

```
PAPERLESS_FILENAME_FORMAT="tags[foo]-tags[3]"
```

If I have not missed anything, I think this would give the same result. Let me know what you think! :slightly_smiling_face:

Reply:
Fixed